### ISSUE TITLE
Error in labRcpp(ncol(lr)) : negative length vectors are not allowed

### ISSUE THREAD

**Issue by grayfall** [tpq/propr#13]
Greetings. I'm experiencing what seems to be a size-dependent bug in `propr`. Here is a reproducible example

```
library(dplyr)
library(propr)
n.samples <- 80
n.features <- 47000
feature.counts <- rpois(n.features*n.samples, lambda=10000) %>% matrix(nrow=n.samples, ncol=n.features)

phs <- propr(feature.counts, metric='phs')
```
The traceback doesn't show anything particularly useful:

```
Error in labRcpp(ncol(lr)) : negative length vectors are not allowed

2. labRcpp(ncol(lr))
1. propr(feature.counts, metric = "phs")
```

Reducing the number of features to 46000 (and anything below that) "solves" the issue.

**Comment by tpq** [comment]
Hey! Thanks for your interest in propr!

Hm, my first guess is that you have run out of RAM. Rcpp will try to make an object size n.features * n.features, which takes up a lot of space. 

Usually this gives an error like "Error: cannot allocate vector of size 3.8 Gb". When I call "labRcpp(47000)" directly... I get the same error as you.

Perhaps see here -- https://github.com/alyssafrazee/polyester/issues/41#issuecomment-249474607 -- indeed, this may be a generic way that R says "no more memory!"

**Comment by tpq** [comment]
If you have some programming skills, I got a very basic script that divides up the propr analysis into a few chunks, then glues those chunks back together at the end. It was intended to be used for parallelization, but could be used to reduce RAM overhead too. I think you'd just need to replace the foreach() call with a basic for-loop.

[propr.dopar.zip](https://github.com/tpq/propr/files/4839304/propr.dopar.zip)

**Comment by grayfall** [comment]
@tpq Thanks for replying. I doubt this is a memory issue. I've been running this on a machine with 0.5TB of RAM, and memory consumption never went past 100GB.

**Comment by tpq** [comment]
Hm. `labRcpp` is a very simple function, written in C++ via Rcpp. Perhaps something about C++ not having enough bits for the integer. Interestingly, log(46000, base = 2) is about 46,000 -- so it looks like C++ might use 32 bits for the integer by default.

I'm not sure if I can fix this quickly, but I will have a deeper think after the weekend.

```
// Function to label a half matrix
// [[Rcpp::export]]
List labRcpp(int nfeats){

  int llt = nfeats * (nfeats - 1) / 2;
  Rcpp::IntegerVector partner(llt);
  Rcpp::IntegerVector pair(llt);
  int counter = 0;

  for(int i = 1; i < nfeats; i++){
    for(int j = 0; j < i; j++){
      partner(counter) = i + 1;
      pair(counter) = j + 1;
      counter += 1;
    }
  }

  return List::create(
    _["Partner"] = partner,
    _["Pair"] = pair
  );
}
```

**Comment by grayfall** [comment]
@tpq Yes, I thought this might be an overflow somewhere in the cpp code. Thanks for taking care of this issue.

**Comment by grayfall** [comment]
@tpq seems like one can get the same error message whilst trying to create a vector with length exceeding a 31-bit number https://stackoverflow.com/a/48676389/3846213

**Comment by grayfall** [comment]
@tpq Sorry for spamming so hard, but I guess my previous comment is closer to the crux of the matter than the memory issue. Here I've tested the out-of-memory assumption by writing a function that should've worked with ncol = 47000. The function does nothing but allocating an integer vector. Assuming 64-bit integers, we are dealing with mere ~8GB of RAM. Yet, we still get the same error. I am not exactly sure, why this is happening, though. Given ncol = 47000, ncol * (ncol - 1) / 2 is two times smaller than the 31 uint vector-size limit.

```
> library(Rcpp)
>
> cppFunction("
+ IntegerVector test(int ncol) {
+     int veclen = ncol * (ncol - 1) / 2;
+     IntegerVector vec(veclen);
+     return vec;
+ }
+ ")
> temp <- test(47000)
Error in test(47000) : negative length vectors are not allowed
```

**Comment by tpq** [comment]
No need to apologize -- thanks heaps for the reproducible example!! I'm actually a bit of a C++ newb, but I'm thinking that if we could replace int with a long int, it might work. It looks like this helps with the test function. I can try something similar with `labRcpp` too

```
library(Rcpp)
cppFunction("IntegerVector test(long int size) {
int veclen = size * (size - 1) / 2;
IntegerVector vec(veclen);
return vec;
}
")

vec <- test(47000)
```

**Comment by grayfall** [comment]
@tpq It appears that `int` can be a 16-bit number. The actual number of bits in an `int` is implementation-specific, but my GCC on Linux has 16-bit `int`s. Consequently, what we've been seeing are overflows

```
> cppFunction("
+ IntegerVector test(int size) {
+     int veclen = size * (size - 1) / 2;
+     Rcout << veclen;
+     IntegerVector vec(veclen);
+     return vec;
+ }
+ ")
> test(47000)
-1043007148Error in test(47000) : negative length vectors are not allowed
```

Considering the numbers involved, `nfeats`, `llt` and `counter` are all overflowing. Moreover, I'm afraid `llt` has been overflowing for other users too (it will overflow for `nfeats` > ~350). You can fix the issue by changing those `int` to `int32_t` (which is more robust than `long`).

**Comment by grayfall** [comment]
@tpq if you don't mind, I can go over and replace all `int` declarations with `int32_t` myself and submit a PR.

**Comment by grayfall** [comment]
@tpq 32 bits aren't enough 

```
cppFunction("
#include <stdint.h>

IntegerVector test(int32_t size) {
    int32_t veclen = size * (size - 1) / 2;
    Rcout << veclen;
    IntegerVector vec(veclen);
    return vec;
}
")

> vec <- test(47000)
-1043007148Error in test(47000) : negative length vectors are not allowed
```

This happens because the arithmetic expression on the RHS of `int32_t veclen = size * (size - 1) / 2;` overflows. I suppose you are not seeing this, because your compiler defines `long int` as a 64-bit integer, rather than a 32-bit one (which my GCC does). The easy fix would be to switch all ints to 64-bits, but this might push memory consumption. The other option is to keep most things 32-bit and carefully review things that should be 64-bit.

**Comment by grayfall** [comment]
@tpq I went over the codebase and refactored all integer arithmetic operations to avoid overflows in intermediate results. This allows us to avoid 64-bit integers. You should still add a warning in `propr` (and other related functions) to let users know there is a upper-limit on the number of features, because due to R's limitation, indexing doesn't work on vanilla vectors and matrices with more than 2^31 - 1 entries.

**Comment by grayfall** [comment]
I have retracted my PR, because the overflow fix brought out another problem with integer division (https://github.com/tpq/propr/issues/15). At this point the only sensible thing I can think of is to switch the argument type in `labRcpp` (and other related functions) to a 64-bit int.

**Comment by grayfall** [comment]
@tpq I am under the impression, that making the package compatible with the number of features I'm interested in will be a lot more complicated than changing a couple of 32-bit arguments to 64-bits. I've found other places that can overflow. For example, 

```
std::vector<int> indexPairs(NumericMatrix & X,
                            const String op = "==",
                            const double ref = 0){

  // Index pairs by operator and reference
  std::vector<int> index;
  int nfeats = X.nrow();
  for(int i = 1; i < nfeats; i++){
    for(int j = 0; j < i; j++){

      if(op == "==" || op == "="){
        if(X(i, j) == ref){
          index.push_back(j * nfeats + i + 1);
        }
...
```
Here `j * nfeats` can easily overflow a 32-bit `int` with large enough `nfeats`.

**Comment by tpq** [comment]
I assume it should be possible to define index as a `long int` too?

```
std::vector<long int> index;
```

I'll try to have a closer look at this soon -- sorry, it's been a busy week!

**Comment by grayfall** [comment]
@tpq I guess there are many other places like this one. And I'm not sure how this will play out in the end. I've already tried changing all ints to `int64_t`, but the package doesn't compile that way.

**Comment by grayfall** [comment]
@tpq my understanding is that R has a very complicated relationship with 64-bit integers. There is no native 64-bit integer type, and R integer vectors are indexed by 32-bit signed ints, which limits their effective size (you can still have larger vectors, but there are indexing problems). I believe that's the reason, why a simple swap of integer type doesn't work in a package level.
